### PR TITLE
lab: initial commit

### PR DESCRIPTION
**Pull request by gireeshpunathil**
first version of the lab material with diagnostic report

### CODE REVIEW

**Comment by gireeshpunathil** [comment]
@mhdawson @richardlau  @boneskull  - please review and let me know what do you think.

**Comment by boneskull** [comment]
so we’re doing this workshop then? originally I had been recruited to help out; do you need me there? had been operating under the assumption I wasn’t going.

**Comment by mhdawson** [comment]
@boneskull, @gireeshpunathil had a workshop accepted and I'd suggested he reach out to you in order to see if we can talk about the tool you are developing as well.  Do you plan to have it released by NodeConf.eu?

**Comment by boneskull** [comment]
I released it yesterday! https://github.com/ibm/report-toolkit 

but not sure if I can make the trip. in lieu of my presence, I can provide some materials to show how to use the tool.

**Comment by boneskull** [comment]
I’ll see if I can go regardless

**Comment by richardlau** [comment]
My inital thoughts reading through the files is who is the workshop aimed at? Most of the items under "CHECK" read as though it's technically validating how the report currently works and is more useful for core developers. e.g. for `cli/fatalerror.js` "No valid Javastack is present in the report" is just how it currently works and not particular useful for the purposes of diagnosing the fatal error.

If the aim was to encourage usage of diagnostics reports maybe it would be more interesting to have sample JS files that didn't have the type of failure in their filenames/comments and then demonstrate how the resulting report indicates what the failure was?

**Comment by gireeshpunathil** [comment]
@richardlau - thanks for the insights!

> No valid Javastack is present in the report

that was an extreme case, which I agree was not much meaningful. It was direct traslation of a test case.

In my experience, it is too much to expect from the lab attendees to directly use the tool or start diagnosing issues. A realistic expectation is a set of beginners into Node.js, and a few Node.js programmers who understand problem symtpoms, but not how to resolve those.

My objective through this workshop is demonstrate the most common production issues represented through sample programs, and show how to diagnose and resolve those issues using the report tool.

Let me take a middle-ground approach here: by splitting the lab into 2 sections; first one showcasing the usage, and second one focussing on problem determination through report, hiding the details.

let me know what do you think!

**Comment by gireeshpunathil** [comment]
@boneskull - thanks for the offer, I would love if you can join me and help in the lab!

**Comment by mhdawson** [comment]
@gireeshpunathil talking with @boneskull it does not sound like he'll be able to make it but should be able to help in advance with any use/integration of the tool that he has released into the lab.

**Comment by gireeshpunathil** [comment]
@boneskull - please take a look, especially `rtk` folder:

 - added scenarios to cover `diff`, `inspect` and `redact`.
 - for inspect, it would have been better to cover more scenarios like `CPU`, `loop` etc. , but I guess the tool still need to cover those? I did not get any output when I applied `rtk inspect` for at least high CPU scenarios.

**Comment by gireeshpunathil** [comment]
I was also unable to use the `transform` functionality effectively. I guess the commands listed in the guide is not working.

**Comment by boneskull** [comment]
I’ll take a look tomorrow. the cpu/memory rules may be disabled by default. I haven’t yet written docs about configuration.

**Comment by boneskull** [comment]
@gireeshpunathil 

> for inspect, it would have been better to cover more scenarios like CPU, loop etc. , but I guess the tool still need to cover those? I did not get any output when I applied rtk inspect for at least high CPU scenarios.

Can you send me the `report.json` you used?  You can also try this to see everything it finds:

```shell
$ rtk inspect report.xxxxxx.json --severity info
```

I'm adding the "memory usage" rule to the default ("recommended") configuration and this will be published in `v0.2.0`.

**Comment by boneskull** [comment]
@gireeshpunathil Please install `report-toolkit@0.2.0` and try that `transform` command again

**Comment by boneskull** [comment]
the talk I gave at Connect Tech the other week, and will give again at Node+JS Interactive is essentially 
 what @richardlau is suggesting.  here are slides, though I can't claim they will be helpful: https://speakerdeck.com/boneskull/diagnostic-reports

File: cli/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+
+In this lab, we learn:
```
Comment (mhdawson) [inline cli/README.md]:
```suggestion
In this lab, we learn how to:
```

File: cli/README.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+
+ - enable diagnostic report
+ - use different flags pertinent to relevant problem symptoms
+ - understand different sections in the report
+ - map it with the control flow in the program that originated it
```
Comment (mhdawson) [inline cli/README.md]: what is `it` ?

**Comment by mhdawson** [comment]
@gireeshpunathil a few questions:

1) you mentioned breaking out into 2 sections? Have you progressed on that?
2) Do you plan for any slides/intro at the start?
3) How long do you think it will take people to get through the existing scenarios?

My initial impression is that it would be good to pull the instructions out into the readme along with a bit more explanation, guidance so that it's easier to see the overall flow and the participants don't just get a list of files they have to open up to see the instructions as comments.

File: cli/README.md (hunk 1 above)
Comment (gireeshpunathil) [reply cli/README.md]: thanks @mhdawson , fixed.

File: cli/README.md (hunk 2 above)
Comment (gireeshpunathil) [reply cli/README.md]: addressed, expanded `it`.

**Comment by gireeshpunathil** [comment]
@mhdawson - thanks.

> you mentioned breaking out into 2 sections? Have you progressed on that?

TBH, I guess that will be too much for the attendees to cover meaningfully, based on their skill level as observed in the past. However, if you two are convinced sufficiently about that (hiding the issue in the program and let the attendee discover that through report), PLMK, I can refactor the content such.

> Do you plan for any slides/intro at the start?

Yes. I plan to spend 15 - 20 minutes of theory:
 - intro to asynchronous event driven programming model
 - common problem types in node.js deployments
 - intro to diagnostic report
 - structure, flow and objective of the lab

let me know what do you think.

> How long do you think it will take people to get through the existing scenarios?

My gut feeling is that an average attendee will be able to complete both `cli` and `cfg` in 2 - 3 hours, but will not be able to complete `rtk` fully. A full hour may be consumed for the theory plus the system setup / node install etc.

> My initial impression is that it would be good to pull the instructions out into the readme along with a bit more explanation, guidance so that it's easier to see the overall flow and the participants don't just get a list of files they have to open up to see the instructions as comments.

Sure, I will come up with one!

thanks again!

**Comment by gireeshpunathil** [comment]
I have been asked by the organizer to provide a link to the lab at the earliest so that they can share it with potential attendees.

I guess I have addressed all the outstanding reviews. Unless there are no more questions / comments, I plan to freeze on the current stuff as is. (by tomorrow)

thanks for all the review comments, @richardlau @boneskull @mhdawson !

**Comment by mhdawson** [comment]
Are you still planning to do:

> My initial impression is that it would be good to pull the instructions out into the readme along with a bit more explanation, guidance so that it's easier to see the overall flow and the participants don't just get a list of files they have to open up to see the instructions as comments.

**Comment by boneskull** [comment]
@gireeshpunathil re the CPU thing:

I didn't like the experience you had using it with Node.js v12.x: it should display _something_; at the very least, it needs to display a warning that it could not compute the CPU usage.  I will need to make sure the severity of that message is set to "warning" and change the default threshold to "warning" as well.  Currently, I think the message was "info" (too low) and the default threshold is "error" (too high).

**Comment by gireeshpunathil** [comment]
@mhdawson - yes, I have a 3 page handout ready, with elaborate instructions. I will pass it to the organizers to get it printed in hard copy - handy to use in the lab.

**Comment by boneskull** [comment]
@gireeshpunathil You should now see a warning by default when running `inspect` on a pre-Node-13 report; upgrade to `report-toolkit@0.2.1`.

image

I'm still not sure this is the right solution, given most reports will probably be generated with pre-Node-13.  What do you think?

**Comment by gireeshpunathil** [comment]
@boneskull - thanks; I will experiment with that a bit, and will share my opinion / observation.

**Comment by gireeshpunathil** [comment]
landed as a85c1ac0088ff3b125eec8599cd7dff0ac8a2565

**Pull request closed without merging**